ui/src/app/views/account-creation/component.tsx

```diff
@@ -0,0 +1,312 @@
+      return;
+    }
+    const request: CreateAccountRequest = {
+      profile: this.state.profile,
+      invitationKey: invitationKey
+    };
+    this.setState({creatingAccount: true});
+    const args: FetchArgs = ProfileApiFetchParamCreator().createAccount(request);
+    fetch(fullUrl(args.url), args.options)
+      .then(handleErrors)
+      .then((response) => response.json)
+      .then((profile) => {
+          this.setState({profile: profile, creatingAccount: false});
+          this.accountCreated = true;
+          setProfile(profile);
+          onAccountCreation();
+        }
+      )
+      .catch(error => {
+        console.log(error);
+        this.setState({creatingAccount: false});
+      });
+  }
+
+  usernameNotValid(): boolean {
+    if (isBlank(this.state.profile.username) || this.state.usernameCheckInProgress) {
+      return false;
+    }
+    return this.isUsernameValidationError();
+  }
+
+  usernameValid(): boolean {
+    if (isBlank(this.state.profile.username) || this.state.usernameCheckInProgress) {
+      return false;
+    }
+    return !this.isUsernameValidationError();
+  }
+
+  isUsernameValidationError(): boolean {
+    return (this.state.usernameConflictError || this.usernameInvalidError());
+  }
+
+  usernameInvalidError(): boolean {
+    const username = this.state.profile.username;
+    if (isBlank(username)) {
+      return false;
+    }
+    if (username.trim().length > 64) {
+      return true;
+    }
+    // Include alphanumeric characters, -'s, _'s, apostrophes, and single .'s in a row.
+    if (username.includes('..') || username.endsWith('.')) {
+      return true;
+    }
+    return !(new RegExp(/^[\w'-][\w.'-]*$/).test(username));
+  }
+
+  usernameChanged(value: string): void {
+    this.updateProfile('username', value);
+    if (this.state.profile.username === '') {
+      return;
+    }
+    this.setState({usernameConflictError: false});
+    // TODO: This should use a debounce, rather than manual setTimeout()s.
+    clearTimeout(this.usernameCheckTimeout);
+    this.setState({usernameCheckInProgress: true});
+    this.usernameCheckTimeout = setTimeout(() => {
+      if (!this.state.profile.username.trim()) {
+        this.setState({usernameCheckInProgress: false});
+        return;
+      }
+      const args: FetchArgs = ProfileApiFetchParamCreator()
+        .isUsernameTaken(this.state.profile.username);
+      fetch(fullUrl(args.url), args.options)
+        .then(handleErrors)
+        .then((response) => response.json())
+        .then((body) => {
+          this.setState({usernameCheckInProgress: false, usernameConflictError: body.isTaken});
+        })
+        .catch((error) => {
+          console.log(error);
+          this.setState({usernameCheckInProgress: false});
+        });
+    }, 300);
+  }
+
+  updateProfile(attribute: string, value: string) {
+    const newProfile = this.state.profile;
+    newProfile[attribute] = value;
+    this.setState({profile: newProfile});
```

Comment:
It's generally a good idea to avoid mutating state if possible. You could use the function form of `setState` along with a `_.set` to implement this function as:
```
this.setState(({profile}) => ({profile: _.set(attribute, value, profile)}));
```

Reply:
So I tried this out, and typescript compilation is complaining on 'attribute' because it's expecting an object but I'm passing a string?

Reply:
nvm. figured it out (order of args is actually: profile, attribute, value)